src/ralph/admin/templates/admin/base.html

```diff
@@ -25,37 +25,40 @@
               <a href="#"><span>Menu</span></a>
             </li>
           </ul>
-          {% cache 86400 header_menu opts.app_label opts.model_name request.user.permissions_hash %}
-          <!-- cache -->
-          {% spaceless %}
-            <section class="top-bar-section">
+          <section class="top-bar-section">
+            <!-- user cache key: {{ request.user.permissions_hash }} -->
+            {% cache 86400 header_menu opts.app_label opts.model_name request.user.permissions_hash %}
+            {% spaceless %}
+            <!-- cache -->
               {% sitetree_menu from "ralph_admin" include "trunk" template "admin/menu_admin.html" %}
-              <ul class="right">
+            {% endspaceless %}
+            <!-- end cache -->
+            {% endcache %}
+            <ul class="right">
+              {% if cl.opts %}
                 <li class="has-form">
                   {% contextual_search_form search_url search_fields cl.opts.verbose_name %}
                 </li>
-                {% block usertools %}
-                  {% if has_permission %}
-                    {% block welcome-msg %}
-                      <li class="has-dropdown not-click">
-                        <a href="">
-                          {% firstof user.get_full_name user.get_username %}
-                        </a>
-                        <ul class="dropdown">
-                          {% block userlinks %}
-                            {% include 'admin/partials/userlinks.html' %}
-                          {% endblock %}
-                        </ul>
-                      </li>
-                    {% endblock %}
-                  {% endif %}
-                {% endblock %}
-                {% block nav-global %}{% endblock %}
-              </ul>
-            </section>
-          {% endspaceless %}
-          <!-- end cache -->
-          {% endcache %}
+              {% endif %}
+              {% block usertools %}
```

Comment:
IMO we don't need because the benefits of caching are negligible but if something goes wrong with cache key (duplicates, empty) then user see own name in the right corner and then user not panic..
BTW ``contextual_search_form`` was cached which is wrong